### PR TITLE
ENH: signal: Add Kaiser-Bessel derived window function

### PR DESCRIPTION
**Pull request by AtsushiSakai**
#### Reference issue

PR #4747

#### What does this implement/fix?

About 6 years ago, the [Kaiser-Bessel derived windows](https://en.wikipedia.org/wiki/Kaiser_window#Kaiser%E2%80%93Bessel-derived_(KBD)_window) function was tried to be added in #4747.
However, it is staled now unfortunately despite the window being commonly used in audio signal processing and audio coding.
So, I added the Kaiser-Bessel derived window function again based on the PR #4747.

### CODE REVIEW

**Review by tupui — COMMENTED** [review]
Thanks @AtsushiSakai for reviving this! Can you add @nils-werner as co-author of at least one commit? I think this is important.

File: scipy/signal/tests/test_windows.py [hunk 1]
```diff
@@ -437,6 +437,38 @@ def test_basic(self):
...
+        assert_allclose(windows.kaiser_bessel_derived(2, beta=np.pi / 2)[:1],
+                        np.sqrt(2) / 2)
+
+        assert_allclose(windows.kaiser_bessel_derived(4, beta=np.pi / 2)[:2],
+                        [0.518562710536, 0.855039598640])
+
+        assert_allclose(windows.kaiser_bessel_derived(6, beta=np.pi / 2)[:3],
+                        [0.436168993154, 0.707106781187, 0.899864772847])
+
+    def test_exceptions(self):
```
Comment (tupui) [inline scipy/signal/tests/test_windows.py:461]: We have to match the error here.

File: scipy/signal/windows/_windows.py [hunk 2]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+    ...             label=f"{alpha=}")
+    >>> plt.legend()
+    >>> plt.grid(True)
+    >>> plt.title("Kaiser-Bessel derived window")
+    >>> plt.ylabel("Amplitude")
+    >>> plt.xlabel("Sample")
+    >>> plt.tight_layout()
+    >>> plt.show()
+    """
+
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1271]:
```suggestion
```

File: scipy/signal/windows/_windows.py [hunk 3]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+    Examples
+    --------
+    Plot the Kaiser-Bessel derived window based on the wikipedia
+    reference [1]_:
+
+    >>> from scipy import signal
+    >>> import matplotlib.pyplot as plt
+    >>> N = 100
+    >>> for alpha in [0.64, 2.55, 7.64, 31.83]:
+    ...     plt.plot(signal.windows.kaiser_bessel_derived(N, alpha),
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1261]: Use the OOP interface of matplotlib

File: scipy/signal/windows/_windows.py [hunk 4]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
     return _truncate(w, needs_trunc)
 
 
+def kaiser_bessel_derived(M, beta, sym=True):
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1217]:
```suggestion
def kaiser_bessel_derived(M, beta, *, sym=True):
```

File: scipy/signal/tests/test_windows.py [hunk 5]
```diff
@@ -437,6 +437,38 @@ def test_basic(self):
...
+        assert_allclose(w, w2)
+
+        # Test for Princen-Bradley condition
+        assert_allclose(w[:M // 2] ** 2 + w[-M // 2:] ** 2, 1.)
+
+        assert_allclose(windows.kaiser_bessel_derived(2, beta=np.pi / 2)[:1],
+                        np.sqrt(2) / 2)
+
+        assert_allclose(windows.kaiser_bessel_derived(4, beta=np.pi / 2)[:2],
+                        [0.518562710536, 0.855039598640])
```
Comment (tupui) [inline scipy/signal/tests/test_windows.py:456]: Where are the values coming from? A docstrings would be helpful here and below.

**Review by tupui — COMMENTED** [review]
This is in good shape 👍 A few suggestions.

File: scipy/signal/tests/test_windows.py (hunk 1 above)
Comment (AtsushiSakai) [reply scipy/signal/tests/test_windows.py:466]: Used `with assert_raises(ValueError, match=msg)`

File: scipy/signal/tests/test_windows.py (hunk 5 above)
Comment (AtsushiSakai) [reply scipy/signal/tests/test_windows.py:461]: Added docs

File: scipy/signal/windows/_windows.py (hunk 3 above)
Comment (AtsushiSakai) [reply scipy/signal/windows/_windows.py]: Used OOP interfaces.

**Comment by AtsushiSakai** [comment]
The CI failure looks unrelated.

File: scipy/signal/windows/_windows.py [hunk 6]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
 
+def kaiser_bessel_derived(M, beta, *, sym=True):
+    """Return a Kaiser-Bessel derived window.
+
+    Parameters
+    ----------
+    M : int
+        Number of points in the output window. If zero or less, an empty
+        array is returned. Note that this window is only defined for an even
+        number of points. It will raise `ValueError` otherwise.
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1225]:
```suggestion
        number of points.
```

File: scipy/signal/tests/test_windows.py [hunk 7]
```diff
@@ -437,6 +437,47 @@ def test_basic(self):
...
+
+        assert_allclose(windows.kaiser_bessel_derived(6, beta=np.pi / 2)[:3],
+                        [0.436168993154, 0.707106781187, 0.899864772847])
+
+    def test_exceptions(self):
+        M = 100
+        # Assert ValueError for odd window length
+        msg = "Kaiser-Bessel Derived windows are only defined for even " \
+              "number of points"
```
Comment (tupui) [inline scipy/signal/tests/test_windows.py:470]: Avoid using this
```suggestion
        msg = ("Kaiser-Bessel Derived windows are only defined for even "
               "number of points")
```

File: scipy/signal/tests/test_windows.py [hunk 8]
```diff
@@ -437,6 +437,47 @@ def test_basic(self):
...
+        with assert_raises(ValueError, match=msg):
+            windows.kaiser_bessel_derived(M + 1, beta=4.)
+
+        # Assert ValueError for non-symmetric setting
+        msg = "Kaiser-Bessel Derived windows are only defined for " \
+              "symmetric shapes"
```
Comment (tupui) [inline scipy/signal/tests/test_windows.py:476]:
```suggestion
        msg = ("Kaiser-Bessel Derived windows are only defined for "
               "symmetric shapes")
```

File: scipy/signal/windows/_windows.py [hunk 9]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+    >>> ax.set_ylabel("Amplitude")
+    >>> ax.set_xlabel("Sample")
+    >>> fig.legend(loc="center")
+    >>> fig.tight_layout()
+    >>> fig.show()
+    """
+    if not sym:
+        raise ValueError(
+            "Kaiser-Bessel Derived windows are only defined for symmetric "
+            "shapes")
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1275]:
```suggestion
        raise ValueError(
            "Kaiser-Bessel Derived windows are only defined for symmetric "
            "shapes"
        )
```

File: scipy/signal/windows/_windows.py [hunk 10]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+        raise ValueError(
+            "Kaiser-Bessel Derived windows are only defined for even number "
+            "of points")
+
+    kaiser_window = kaiser(M // 2 + 1, beta)
+    csum = np.cumsum(kaiser_window)
+    half_window = np.sqrt(csum[:-1] / csum[-1])
+    w = np.zeros(M)
+    w[:M//2] = half_window
+    w[-M//2:] = half_window[::-1]
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1288]: I would just do this. I also think it's clearer.
```suggestion
    w = np.concatenate((half_window, half_window[::-1]), axis=0)
```

File: scipy/signal/windows/_windows.py [hunk 11]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+    >>> for alpha in [0.64, 2.55, 7.64, 31.83]:
+    ...     ax.plot(signal.windows.kaiser_bessel_derived(N, alpha),
+    ...             label=f"{alpha=}")
+    >>> ax.grid(True)
+    >>> ax.set_title("Kaiser-Bessel derived window")
+    >>> ax.set_ylabel("Amplitude")
+    >>> ax.set_xlabel("Sample")
+    >>> fig.legend(loc="center")
+    >>> fig.tight_layout()
+    >>> fig.show()
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1270]: We are passing `alpha` with the same values as in the wiki. But the function is parametrized using `beta` so it should be `np.pi * alpha`.

Looking at the wiki, I would also use the same axis (add 0.707 on y-axis and n, 2N-1 on x-axis)

**Review by tupui — COMMENTED** [review]
From the formula, implementation LGTM. Some last suggestions.

File: scipy/signal/windows/_windows.py [hunk 12]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+        When True (default), generates a symmetric window, for use in filter
+        design. It will raise `ValueError` otherwise.
+
+    Returns
+    -------
+    w : ndarray
+        The window, normalized to fulfil the Princen-Bradley condition.
+
+    Notes
+    -----
```
Comment (tupui) [inline scipy/signal/windows/_windows.py:1240]:
```suggestion
        The window, normalized to fulfil the Princen-Bradley condition.

    See Also
    --------
    kaiser

    Notes
    -----
```

File: scipy/signal/windows/_windows.py [hunk 13]
```diff
@@ -1213,6 +1214,81 @@ def kaiser(M, beta, sym=True):
...
+    It is designed to be suitable for use with the modified discrete cosine
+    transform (MDCT) and is mainly used in audio signal processing and
+    audio coding.
+
+    .. versionadded:: 1.9.0
+
+    References
+    ----------
+    .. [1] Wikipedia, "Kaiser window",
```
[the remark this answers is not in the document]
Comment (AtsushiSakai) [reply scipy/signal/windows/_windows.py]: OK. I added it as the first reference.

File: scipy/signal/windows/_windows.py (hunk 11 above)
Comment (AtsushiSakai) [reply scipy/signal/windows/_windows.py]: Indeed. I improved the example figure. What do you think?
SS 2022-01-05 at 13 29 49

Comment (tupui) [reply scipy/signal/windows/_windows.py]: This looks great thanks!

**Review by tupui — APPROVED** [review]
LGTM, thanks again @AtsushiSakai for reviving this and getting it to the finish line. I will merge now. Can you add a note in the release notes?

**Comment by AtsushiSakai** [comment]
Thank you for reviewing and merging. I updated the 1.9.0 release note: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.9.0

**Comment by tupui** [comment]
Thanks @AtsushiSakai

**Pull request merged**